### DISCUSSION TITLE
Estimated/intended release date for 4.0.0 final?

### DISCUSSION THREAD

**Discussion by IanKemp** [riok/mapperly#1446] [Q&A] [answered]
3.6.0 was released on June 18, #1335 was merged 2 days later and I would really really like to use it, but obviously this isn't possible in my corporate environment as 4.0.0 is still in beta. Is there any estimated or intended date for the non-beta release of the latter version?

**Comment by latonz** [comment, marked as answer]
It may still take some time for 4.0 to be released.

Almost all (breaking) changes we want to land in 4.0 are now open as PRs, but the feedback loops and the further preview releases to ensure the stability will still take some time…

**Reply by HaydnDias** [reply]
Also wondering when 4.0 will be released, I have a need for the changes in #1335 too, are you any clearer on a potential release time?


Alternatively would you consider any of the preview v4.0.0 versions stable, but just waiting for further breaking changes?

**Reply by latonz** [reply]
The only breaking changes that are not yet ready that we would like to include in v4.0 are #1384 (not a breaking API change, but may change implementation details significantly) and #1483 (a breaking API change, but should not require source code changes by most users).
I cannot comment on how much time it will take to complete these features as this is a free open source project and we do not make commitments for such projects, also the features mentioned are both contributed by community members 😊. Hope you understand.

We currently have no intention of breaking the v4.0 API again (except for #1483, which should affect almost no users), unless there is an unexpected bug or a lot of feedback on the new features that would require breaking the API again. Other than this, `4.0.0-next.3` should be "stable".

**Reply by HaydnDias** [reply]
Thank you for your response, completely understand and appreciate where you're coming from, thank you for all the hard work you do on this!
